### PR TITLE
fix: added Post Confirmation trigger permission from auth to Custom trigger template

### PR DESCRIPTION
**Pull request by akshbhu**
#### Description of how you validated changes

* Manuel tested and e2e tests also added

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-e2e-tests/src/__tests__/auth_2.test.ts [hunk 1]
```diff
@@ -97,7 +99,7 @@ describe('amplify add auth...', () => {
     expect(lambdaFunction.Configuration.Environment.Variables.GROUP).toEqual('mygroup');
   });
 
-  it('...should allow the user to add auth via API category, with a trigger and function dependsOn API', async () => {
+  it.only('...should allow the user to add auth via API category, with a trigger and function dependsOn API', async () => {
```
Comment (cjihrig) [inline packages/amplify-e2e-tests/src/__tests__/auth_2.test.ts:102]: Did you mean to leave this? FYI, you can use Jest's `-t` flag to only run specific tests.

File: packages/amplify-e2e-core/src/utils/sdk-calls.ts [hunk 2]
```diff
@@ -144,6 +144,32 @@ export const getUserPoolClients = async (userPoolId: string, clientIds: string[]
...
+      ValidationData: [
+        {
+          Name: 'demoKey' /* required */,
+          Value: 'demoValue',
+        },
+      ],
+    };
+    const res = await provider.adminCreateUser(params);
+  } catch (e) {
+    console.log(e);
```
Comment (cjihrig) [inline packages/amplify-e2e-core/src/utils/sdk-calls.ts:169]: Why don't we want to surface the error here in our tests?

File: packages/amplify-category-auth/provider-utils/awscloudformation/triggers/PostConfirmation/cloudformation-templates/PostConfirmation.json.ejs [hunk 3]
```diff
@@ -196,6 +196,10 @@
                 "Ref": "LambdaExecutionRole"
             }
         },
+        "LambdaExecutionRoleArn": {
+            "Value": {"Fn::GetAtt": ["LambdaExecutionRole", "Arn"]}
```
Comment (attilah) [inline packages/amplify-category-auth/provider-utils/awscloudformation/triggers/PostConfirmation/cloudformation-templates/PostConfirmation.json.ejs:200]: Could you reformat like the expression above, into multiple lines with proper ident?

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts [hunk 4]
```diff
@@ -45,6 +46,13 @@ export class CustomResourceAuthStack extends cdk.Stack {
         type: 'String',
       });
       createPermissionToInvokeLambda(this, fnName, userpoolArn, config);
+      const permission = props.permissions.find(permission => config.triggerType === permission.trigger);
```
Comment (attilah) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts:49]: Does `props.permissions` always have a value?

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts [hunk 5]
```diff
@@ -61,9 +69,15 @@ export async function generateNestedAuthTriggerTemplate(context: $TSContext, cat
   const cfnFileName = 'auth-trigger-cloudformation-template.json';
   const targetDir = path.join(pathManager.getBackendDirPath(), category, request.resourceName!);
   const authTriggerCfnFilePath = path.join(targetDir, cfnFileName);
-  const { authTriggerConnections } = request;
+  const { authTriggerConnections, permissions } = request;
   if (authTriggerConnections) {
-    const cfnObject = await createCustomResourceforAuthTrigger(context, JSON.parse(authTriggerConnections));
+    const cfnObject = await createCustomResourceforAuthTrigger(
+      context,
+      JSON.parse(authTriggerConnections),
```
Comment (attilah) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts:76]: Use `JSONUtilities`

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts [hunk 6]
```diff
@@ -61,9 +69,15 @@ export async function generateNestedAuthTriggerTemplate(context: $TSContext, cat
...
+      permissions!.map(i => JSONUtilities.parse(i)),
```
Comment (attilah) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts:77]: Using `!` is for sure? it can never be `undefined`? Alsp why parse the `permissions` array per member, why not just once the whole array?

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts [hunk 7]
```diff
@@ -75,10 +89,15 @@ export async function generateNestedAuthTriggerTemplate(context: $TSContext, cat
...
-async function createCustomResourceforAuthTrigger(context: any, authTriggerConnections: AuthTriggerConnection[]) {
+async function createCustomResourceforAuthTrigger(
+  context: any,
+  authTriggerConnections: AuthTriggerConnection[],
+  permissions: AuthTriggerPermissions[],
+) {
   const stack = new CustomResourceAuthStack(undefined as any, 'Amplify', {
     description: 'Custom Resource stack for Auth Trigger created using Amplify CLI',
     authTriggerConnections: authTriggerConnections,
+    permissions: permissions,
```
Comment (attilah) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts:100]: no need for `authTriggerConnections` and `permissions` after the colon as the names are the same.

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts [hunk 8]
```diff
@@ -121,3 +140,24 @@ function createPermissionToInvokeLambda(
...
+  roleArn: cdk.CfnParameter,
+  permissions: AuthTriggerPermissions,
+  userpoolArn: cdk.CfnParameter,
+) {
+  const myRole = iam.Role.fromRoleArn(stack, 'LambdaExecutionRole', roleArn.valueAsString);
+  new iam.Policy(stack, `${fnName}${permissions.trigger}${permissions.policyName}`, {
+    policyName: permissions.policyName,
+    statements: [
+      new iam.PolicyStatement({
+        effect: permissions.effect === iam.Effect.ALLOW ? iam.Effect.ALLOW : iam.Effect.DENY,
```
Comment (attilah) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts:156]: is permissions.effect is already allow or deny, why do we need this here?

File: packages/amplify-e2e-core/src/categories/auth.ts [hunk 9]
```diff
@@ -1579,3 +1579,58 @@ export function updateAuthWithoutTrigger(cwd: string, settings: any): Promise<vo
       });
   });
 }
+
+export function updateAuthWithGroupTrigger(cwd: string, settings: any): Promise<void> {
+  return new Promise((resolve, reject) => {
+    spawn(getCLIPath(), ['update', 'auth'], { cwd, stripColors: true })
+      .wait('What do you want to do?')
+      .send(KEY_DOWN_ARROW)
```
Comment (attilah) [inline packages/amplify-e2e-core/src/categories/auth.ts:1587]: we have a dedicated method for this with repetitions support, also this should be a `SingleSelect`,  not arrow key usage for better control.

File: packages/amplify-e2e-core/src/categories/auth.ts [hunk 10]
```diff
@@ -1579,3 +1579,58 @@ export function updateAuthWithoutTrigger(cwd: string, settings: any): Promise<vo
...
+      .send(KEY_DOWN_ARROW)
+      .sendCarriageReturn()
+      .wait('Select the authentication/authorization services that you want to use:')
+      .sendCarriageReturn()
```
Comment (attilah) [inline packages/amplify-e2e-core/src/categories/auth.ts:1591]: Don't propagate the bad behavior we are having in existing tests to rely on option defaults, can lead to hard to debug errors, use explicit options and yes/no answers.

File: packages/amplify-e2e-tests/src/__tests__/auth_2.test.ts [hunk 11]
```diff
@@ -120,10 +122,18 @@ describe('amplify add auth...', () => {
     const functionName = `${authKey}PostConfirmation-integtest`;
     const authMeta = meta.auth[authKey];
     const id = authMeta.output.UserPoolId;
-    const userPool = await getUserPool(id, meta.providers.awscloudformation.Region);
+    const region = meta.providers.awscloudformation.Region;
```
Comment (attilah) [inline packages/amplify-e2e-tests/src/__tests__/auth_2.test.ts:125]: enhance the test for deny cases with different user and group assignment for better coverage.

File: packages/amplify-e2e-tests/src/__tests__/auth_2.test.ts [hunk 12]
```diff
@@ -120,10 +122,18 @@ describe('amplify add auth...', () => {
...
+    const userPool = await getUserPool(id, region);
     const clientIds = [authMeta.output.AppClientIDWeb, authMeta.output.AppClientID];
-    const clients = await getUserPoolClients(id, clientIds, meta.providers.awscloudformation.Region);
-    const lambdaFunction = await getLambdaFunction(functionName, meta.providers.awscloudformation.Region);
+    const clients = await getUserPoolClients(id, clientIds, region);
+    await addUserToUserPool(id, region);
+    const lambdaFunction = await getLambdaFunction(functionName, region);
```
Comment (attilah) [inline packages/amplify-e2e-tests/src/__tests__/auth_2.test.ts:130]: Are we invoking the PostConfirmation trigger lambda directly instead of testing it through Cognito to see if it is properly configured and working as expected? If that's the case, we need to test it end-to-end, not in isolation like this.

File: yarn.lock [hunk 13]
```diff
@@ -9723,7 +9723,7 @@ camel-case@4.1.1, camel-case@^4.1.1:
     pascal-case "^3.1.1"
```
Comment (attilah) [inline yarn.lock:9723]: Unrelated changes, remove `yarn.lock` from the PR.

**Review by attilah — CHANGES_REQUESTED** [review]

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts (hunk 6 above)
Comment (akshbhu) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts]: Removed !.
 Parsing is like this as data is sent in this format previously from auth walkthrough

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts (hunk 8 above)
Comment (akshbhu) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts:159]: These policies are coming from AuthTriggers resource file ```AuthTriggerName.map.json```. Based on this value I am toggling permissions in this file.

File: packages/amplify-e2e-core/src/categories/auth.ts (hunk 10 above)
Comment (akshbhu) [reply packages/amplify-e2e-core/src/categories/auth.ts]: same not needed

File: packages/amplify-e2e-tests/src/__tests__/auth_2.test.ts (hunk 12 above)
Comment (akshbhu) [reply packages/amplify-e2e-tests/src/__tests__/auth_2.test.ts]: I tried with ```AdminConfirmSingup``` Call which calls PostConfirmation Auth Trigger , But there is no way currently as When I tried invoking though USerPool, got following errors : 

AdminCreateUser -> Admin ConfirmSignUp -> error: user state is FORCE_PASSWD_CHANGE(NOT Authorized)
AdminCreateUser -> AdminSetPassword ->  Admin ConfirmSignUp -> error: user state is CONFIRMED(NOT Authorized)

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts (hunk 4 above)
Comment (akshbhu) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/utils/generate-auth-trigger-template.ts:49]: Nice catch. Updated . Now this is checked in generateNestedAuthTriggerTemplate() Fn.

File: packages/amplify-e2e-tests/src/__tests__/auth_2.test.ts (hunk 12 above)
Comment (attilah) [reply packages/amplify-e2e-tests/src/__tests__/auth_2.test.ts]: We are creating and confirming users in the graphql e2e tests by using AmplifyJS calls, no need to admin confirm it, use the same method that client applications will use.

**Comment by Kirnos** [comment]
Any update on merging this fix? This is a very annoying regression that needs to be fixed ASAP. Thanks.

**Comment by akshbhu** [comment]
closing in favour of https://github.com/aws-amplify/amplify-cli/pull/11482
